### PR TITLE
tests: generating/getting configuration deployments

### PR DESCRIPTION
**Pull request by mchalski**
https://tracker.mender.io/browse/MEN-4428

---

overall very good results, no bugs found.

I marked some bits I had doubts about as GOTCHAs - Alf could you provide some input?
I'm probably mistaken about some assumptions. when we clear this up I'll remove them from code.

### CODE REVIEW

File: tests/tests/test_configuration_deployments.py [hunk 1]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+        # no x-men-expire
+        uri_no_exp = uri.replace("&x-men-expire=", "")
+        r = requests.get(
+            uri_no_exp,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
+            verify=False,
+        )
+        assert r.status_code == 400
+
+        # GOTCHA: uri for one device can be accessed by a diff device from same tenant (jwt)
```
Comment (mchalski) [inline tests/tests/test_configuration_deployments.py:314]: self-comment:
@alfrunes these are some scenarios that surprised me.
it's all around authentication of the download endpoint.
let me know what you think!

Comment (alfrunes) [reply tests/tests/test_configuration_deployments.py:314]: We can't use the authorization middleware for this endpoint, thus we have to pass all the parameters in the signed URL.
Note that this is also the case for S3 signed URLs.

Comment (alfrunes) [reply tests/tests/test_configuration_deployments.py:314]: This reminds me I forgot to add the exception to the Træfik gateway: https://github.com/mendersoftware/integration/pull/1198

File: tests/tests/test_configuration_deployments.py [hunk 2]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+
+        assert nextdep.artifact["artifact_name"] == test_set["name"]
+        assert nextdep.artifact["source"]["uri"] is not None
+        assert nextdep.artifact["source"]["expire"] is not None
+        assert nextdep.artifact["device_types_compatible"] == [test_set["dev_type"]]
+
+        # get/verify download contents
+        r = requests.get(
+            nextdep.artifact["source"]["uri"],
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:196]: Authorization header is usually not present when calling this URI.
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 3]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+        )
+        assert r.status_code == 403
+
+        # wrong dev type (signature error)
+        other_dev = Device()
+        other_dev.device_type = "foo"
+        uri_bad_devtype = uri.replace(dev.device_type, other_dev.device_type)
+        r = requests.get(
+            uri_bad_devtype,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:271]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 4]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+            verify=False,
+        )
+        assert r.status_code == 403
+
+        # wrong dev id (signature error)
+        uri_bad_devid = uri.replace(dev.devid, other_dev.devid)
+        r = requests.get(
+            uri_bad_devid,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:280]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 5]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+        )
+        assert r.status_code == 403
+
+        # wrong x-men-signature
+        uri_bad_sig = uri.replace(
+            qs["x-men-signature"][0], "mftJRzBafnvMXhmMBH3THQertiEk0dZKP075bjBKccc"
+        )
+        r = requests.get(
+            uri_bad_sig,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:291]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 6]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+            verify=False,
+        )
+        assert r.status_code == 403
+
+        # no x-men-signature
+        uri_no_sig = uri.replace("&x-men-signature=", "")
+        r = requests.get(
+            uri_no_sig,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:300]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 7]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+            verify=False,
+        )
+        assert r.status_code == 400
+
+        # no x-men-expire
+        uri_no_exp = uri.replace("&x-men-expire=", "")
+        r = requests.get(
+            uri_no_exp,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:309]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 8]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+
+        # wrong tenant in url (signature error)
+        other_tenant_id = str(ObjectId())
+        _, r = api_client_int.create_tenant(other_tenant_id)
+        assert r.status_code == 201
+
+        uri_bad_tenant = uri.replace(tenant_id, other_tenant_id)
+        r = requests.get(
+            uri_bad_tenant,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:260]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 9]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+        uri = nextdep.artifact["source"]["uri"]
+        qs = parse_qs(urlparse(uri).query)
+
+        # now break the url in various ways
+
+        # wrong deployment (signature error)
+        uri_bad_depl = uri.replace(deployment_id, str(uuid4()))
+        r = requests.get(
+            uri_bad_depl,
+            headers={"Authorization": "Bearer " + dev.fake_token_mt(tenant_id)},
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:247]:
```suggestion
```

File: tests/tests/test_configuration_deployments.py [hunk 10]
```diff
@@ -132,3 +134,296 @@ def test_fail_duplicate_deployment(self, api_client_int, clean_db):
...
+        start, stop = stdout.index("{", metapos), stdout.index("}", metapos)
+        assert json.loads(config) == json.loads(stdout[start : stop + 1])
+
+        # provides
+        assert (
+            "Provides: data-partition.mender-configure.version: {}".format(name)
+            in stdout
+        )
+        assert 'Clears Provides: ["data-partition.mender-configure.*"]' in stdout
+        assert "Depends: Nothing" in stdout
```
Comment (alfrunes) [inline tests/tests/test_configuration_deployments.py:355]: Nice! :+1:

**Review by alfrunes — CHANGES_REQUESTED** [review]
The acceptance tests covers all the cases that I was hoping for, well done! The only thing I could pick on is the authorization header is usually not present in the request when calling these "external" download URIs as the URL is self-contained. I would like to see at least one request where this is the case to verify that the Authorization header is actually not required.

File: tests/tests/test_configuration_deployments.py (hunk 1 above)
Comment (mchalski) [reply tests/tests/test_configuration_deployments.py:314]: ok, I guess it's clear now from the other comments - auth is not here by design, hence all questions are moot.

File: tests/tests/test_configuration_deployments.py (hunk 2 above)
Comment (mchalski) [reply tests/tests/test_configuration_deployments.py]: ok, I decided to remove Auth from all invocations.

**Review by alfrunes — APPROVED** [review]

**Pull request merged**